pkg/policy/cell/policy_importer.go

```diff
@@ -0,0 +1,421 @@
+
+		} else {
+			// Standard case: there is an owning resource.
+			// Track the complete set of per-prefix resources.
+			// We want empty sets here!
+			toAllocate[upd.Resource] = prefixes
+			prefixSource[upd.Resource] = upd.Source
+		}
+	}
+
+	// Now that we know the exact set of prefixes for each resource, determine ipcache update.
+	// Note that we did this step separately, as we could have batched multiple updates
+	// for the same resource (i.e A1, A2, A3 -> only A3 matters)
+	//
+	// We elide updates when the prefix already has an entry for the given resource.
+	var ipcUpdates []ipcache.MU
+	for resource, newPrefixes := range toAllocate {
+		// For anonymous prefixes, just allocate, don't bookkeep
+		if resource == ResourceIDAnonymous {
+			for _, prefix := range newPrefixes {
+				ipcUpdates = append(ipcUpdates, ipcache.MU{
+					Prefix:   prefix,
+					Source:   prefixSource[resource],
+					Resource: resource,
+					Metadata: []ipcache.IPMetadata{labels.GetCIDRLabels(prefix)},
+				})
+			}
+			continue
+		}
+
+		// otherwise, update bookkeeping and determine diff
+		oldPrefixes := i.prefixesByResource[resource]
+
+		// No prefixes for this resource: clear entry, prune all old prefixes.
+		if len(newPrefixes) == 0 {
+			delete(i.prefixesByResource, resource)
+			if len(oldPrefixes) > 0 {
+				toPrune[resource] = oldPrefixes
+				continue
+			}
+		}
+
+		// Otherwise, update bookkeeping, upsert any net-new prefixes.
+		i.prefixesByResource[resource] = newPrefixes
+		oldSet := set.NewSet(oldPrefixes...)
+		for _, prefix := range newPrefixes {
+			// If we already allocated this prefix, remove it from the prune list
+			// and don't bother upserting it again in to the ipcache
+			if oldSet.Remove(prefix) {
+				continue
+			}
+
+			ipcUpdates = append(ipcUpdates, ipcache.MU{
+				Prefix:   prefix,
+				Source:   prefixSource[resource],
+				Resource: resource,
+				Metadata: []ipcache.IPMetadata{labels.GetCIDRLabels(prefix)},
+			})
+		}
+		if oldSet.Len() > 0 {
+			toPrune[resource] = oldSet.AsSlice()
+		}
+	}
+
+	// Batch the set of updates to the ipcache.
+	if len(ipcUpdates) > 0 {
+		i.log.Info("inserting ipcache metadata for CIDR prefixes from policy", logfields.Count, len(ipcUpdates))
+		nextIPCRev := i.ipc.UpsertMetadataBatch(ipcUpdates...)
+
+		// If the ipcache has already started, then we should wait for our update to commit.
+		// However, no sense in waiting if the agent is still starting up.
+		// For resilience purposes, we will wait a maximum of 10 seconds for the ipcache to make progress.
+		if nextIPCRev > 1 {
+			i.log.Debug("Waiting up to 10 seconds for ipcache to upsert policy prefix metadata")
+			updCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
+			if err := i.ipc.WaitForRevision(updCtx, nextIPCRev); err != nil {
+				i.log.Warn("Timed out waiting for ipcache to allocate identities for prefixes while consuming policy updates. This may cause policy drops!")
+			}
+			cancel()
```

Comment:
What happens if we legitimately exceed 10s here? Why do we need to actually time box this logic?

Reply:
So, the previous code never waited for ipcache to update prefixes, which could lead to drops if the policy is applied before the ipcache finished. In the interest of correctness, we should wait.

However, out of paranoia, I added a time limit here just in case we somehow cause a deadlock. It will emit a warning and cause CI to fail.